### ISSUE TITLE
Define variable shape at restore/load, allow direct restoring of variables prior to calling __build__ (non-lazy variable loading from checkpoint)

### ISSUE THREAD

**Issue by mjlbach** [tensorflow/tensorflow#35685]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.0
- Are you willing to contribute it (Yes/No): Yes (if necessary)

**Describe the feature and the current behavior/state.**

Typically for RL, I use tensorflow in combination with Ray, where a remote agent collects data using a policy model, and the episodes are fed to a tf-agents EpisodicReplayBuffer in a second train-loop process. A tf.dataset is used to feed the main model training on this thread, which is then serialized via get_weights and sent back to the remote actor via a ray remote call to set_weights.

Currently I use tf.Checkpoint.restore.expect_partial() to restore my subclassed tf.keras.Model. In practice, this is extremely convoluted on initialization:
* I make a forward pass through my remote Agent on random policy (since I have not initialized the weights), retrieve this forward pass, and send it to the train loop
* I make a second forward pass through my train loop (again, randomly initialized network still), this has the side effect of initializing the network, allowing my checkpoint to restore the variables in this model
* I call get_weights on the train_loop model, and pass the weights to my remote model via set_weights.
* I can now collect a full replay buffer with my Agent on policy
* I can now run the train loop

Maybe I'm missing something, but most of my labmates are also confused by what the best practice for this currently is.

**Will this change the current api? How?**
I'm not sure what the best approach is. Some ideas for discussion? 

* Allow set_weights to define the shapes and names of tf.Variables, such that calls to __build__ in the tf.keras.Model on first run will align with the initialized tf.Variables.
* Allow tf.keras.model.load_weights() to set the weights immediately rather than waiting for first call to build()

**Who will benefit with this feature?**
RL community 
**Any Other info.**

**Comment by ebrevdo** [comment]
This is partly a question for the TF-Agents repo, and partly for Keras (CC @fchollet).  On the TF-Agents side, it's true you need to do a forward pass to create the appropriate Variables.  The easiest way to do this is to take your TF-Agents network objects, and call net.create_variables() on each of them (you may want to pass training=True to this call in case you have batch norm or dropout).

That takes care of feeding appropriate random inputs.  If that doesn't work, you should file a bug in tensorflow/agents instead of this repo.

Regarding Keras models, we rely on the lower level Keras Network class to represent computation in TF-Agents, so I can't help much there.  I hope @fchollet can discuss your ideas on variable initialization.

**Comment by mjlbach** [comment]
@ebrevdo Thanks for the comments! Just to clarify, the *only* thing I'm using from tf-agents is the EpisodicReplayBuffer class, I'm constructing my own tf.keras model using the low-level apis.

**Comment by Venkat6871** [comment]
Hi @mjlbach  ,
Thank you so much for filing this feature request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.To avoid duplicate work, please leave a comment here if you decide to take it on!
Thank you again for your interest and we look forward to seeing your contributions!
